### PR TITLE
bin/magento must return non zero on exceptions

### PR DESCRIPTION
**Pull request by BlackIkeEagle**
When you use bin/magento in a deployment flow it must return a non zero
exitcode when there was an error or an exception. Now the exceptions
caught and as a result the exitcode is 0. In a script we assume exit
code 0 to be all clear everything went fine.

So we must fix this here.

Signed-off-by: BlackEagle ike.devolder@gmail.com

### CODE REVIEW

**Comment by BlackIkeEagle** [comment]
@vancoz I've added more things that are related to incorrect exit codes, these are open for discussion

**Comment by BlackIkeEagle** [comment]
I'll fix the failing tests probably later today

**Comment by BlackIkeEagle** [comment]
@vancoz Could we discuss here if the changes I've done are good for you. Maybe it can be taken further and instead of returning non-zero from the execute methods we could just throw exceptions instead.

**Comment by BlackIkeEagle** [comment]
Sorry for the delay, I got it sorted out now.

**Comment by vancoz** [comment]
Hi @BlackIkeEagle, thank you for contribution, and sorry for delay.
Can you please use constants for exit codes? Thank you.

**Comment by erikhansen** [comment]
@BlackIkeEagle Did you see the latest feedback from @vancoz about _"Can you please use constants for exit codes"_? Are you going to be able to work on that small change?

Thanks for working on this fix, as it solved PR https://github.com/magento/magento2/issues/3060 that myself and others have wanted to see fix.

**Comment by BlackIkeEagle** [comment]
@erikhansen Sorry I have seen the comment by @vancoz but I did not yet get to it to update this PR.

**Comment by erikhansen** [comment]
@BlackIkeEagle It looks like the latest CI tests failed. I saw other tests that were failing for the same reason, so it looks like it was an issue with code from develop. Also, your branch now has conflicts with develop. Can you merge develop into your branch and fix any merge conflicts? Hopefully that will result in successful CI tests.

**Comment by BlackIkeEagle** [comment]
@erikhansen Ok I'll rebase again :)

**Comment by BlackIkeEagle** [comment]
pff is there any way for magento to use something more stable than travis please

there are almost every time issues with stuff that is rebooting and whatever

**Comment by erikhansen** [comment]
@vancoz It looks like the Travis CI build randomly failed. Can you restart the tests?

File: bin/magento [hunk 1]
```diff
@@ -32,4 +32,5 @@ try {
         echo "\n\n";
         $e = $e->getPrevious();
     }
+    exit(1);
```
Comment (vancoz) [inline bin/magento]: Can you please use constant here?

Comment (BlackIkeEagle) [inline bin/magento]: thanks for spotting that, I completely missed it

**Comment by BlackIkeEagle** [comment]
Fingers crossed travis does not reboot anything now :)

**Comment by BlackIkeEagle** [comment]
please help me here:

integration 1 php 5.5

```
The system is going down for power off NOW!
FEPHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY000] [2002] Connection refused' in
```

integration 1 php 5.6

```
The system is going down for power off NOW!
EPHP Warning:  PDOStatement::execute(): MySQL server has gone away in
```

integration 2 php 7

```
The system is going down for power off NOW!
EPHP Fatal error:  Uncaught PDOException: SQLSTATE[HY000] [2002] Connection refused in
```

integration integrity php 7

```
The system is going down for power off NOW!
Exception: Starting Magento installation:
....
  [Exception]
  Warning: PDOStatement::execute(): MySQL server has gone away in /home/travi
  s/build/magento/magento2/vendor/magento/zendframework1/library/Zend/Db/Stat
  ement/Pdo.php on line 228
```

static:

I don't think the issue is caused by my changes so fixing it here would be out-of-scope unless you guys say 'just fix it here :)'

**Comment by BlackIkeEagle** [comment]
I'm really tired of this stupid failures with travis

**Comment by Vinai** [comment]
Restarted the build.

**Comment by BlackIkeEagle** [comment]
@Vinai thanks, but again there is such an issue:

The system is going down for power off NOW!

**Comment by davidalger** [comment]
> I'm really tired of this stupid failures with travis

@BlackIkeEagle Does it help at all to know others (like myself) are pretty sick of it too? :) Seems to have all started when Travis started shifting their infrastructure over to GCE, and unfortunately the only option we've got is to restart the failed builds.

**Comment by BlackIkeEagle** [comment]
@davidalger I did not at all assume I was alone with my annoyance. So we can conclude if we cumulate all 'failed' builds that is basically succeeds :)

**Comment by mazhalai** [comment]
@BlackIkeEagle there are some static failures specific to your changes. Please see: https://travis-ci.org/magento/magento2/jobs/125264905#L622

**Comment by BlackIkeEagle** [comment]
@mazhalai yeah sorry I already noticed but did not yet have the time to update the PR

**Comment by BlackIkeEagle** [comment]
@mazhalai I checked again, and if you want less dependencies the code must be refactored. Its not like the new 'dependency' will make it more messy.

**Comment by BlackIkeEagle** [comment]
@mazhalai do you want me to add PHPMD ignores or something like that

**Comment by mazhalai** [comment]
We do now allow any _new_ suppress warning, I believe there is a task/story in backlog to remove all existing ones. Yes this will need to be refactored.

**Comment by piotrekkaminski** [comment]
Internal issue id: MAGETWO-52458, reported as well by @LeeSaferite

**Comment by vancoz** [comment]
@BlackIkeEagle  thank you for your contribution, you do not have to do this refactoring, because your changes is the reason of this issue. We going to merge in your changes and we will take care about refactoring.

**Comment by BlackIkeEagle** [comment]
@vancoz thanks

**Comment by vancoz** [comment]
MAGETWO-52000

File: app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php [hunk 2]
```diff
@@ -79,7 +79,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $productIds = $productCollection->getAllIds();
         if (!count($productIds)) {
             $output->writeln("<info>No product images to resize</info>");
-            return;
+            // we must have an exit code higher than zero to indicate something was wrong
+            return \Magento\Framework\Console\Cli::RETURN_FAILURE;
```
Comment (LeeSaferite) [inline app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php:83]: Is this really an error?

File: app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php [hunk 3]
```diff
@@ -99,7 +100,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
         } catch (\Exception $e) {
             $output->writeln("<error>{$e->getMessage()}</error>");
-            return;
+            // we must have an exit code higher than zero to indicate something was wrong
+            return \Magento\Framework\Console\Cli::RETURN_FAILURE;
         }
 
         $output->write("\n");
```
Comment (LeeSaferite) [inline app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php:107]: Seems like you are missing an explicit success return here

File: app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php (hunk 2 above)
Comment (BlackIkeEagle) [inline app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php:83]: @LeeSaferite I rather would have it as an error so you can choose to ignore it in your 'external' tooling

File: app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php (hunk 3 above)
Comment (BlackIkeEagle) [inline app/code/Magento/Catalog/Console/Command/ImagesResizeCommand.php:107]: @LeeSaferite nice catch, I'll update it, thanks.

**Comment by mazhalai** [comment]
@BlackIkeEagle this PR has been merged. It was combined with other logging related issues, hence PR was not automatically closed. Thank you very much for your contribution!

**Comment by BlackIkeEagle** [comment]
@mazhalai superb, thanks

File: app/code/Magento/Deploy/Console/Command/SetModeCommand.php [hunk 4]
```diff
@@ -109,7 +109,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
             if ($output->getVerbosity() >= OutputInterface::VERBOSITY_VERBOSE) {
                 $output->writeln($e->getTraceAsString());
             }
-            return;
+            // we must have an exit code higher than zero to indicate something was wrong
+            return \Magento\Framework\Console\Cli::RETURN_FAILURE;
         }
     }
```
Comment (mc388) [inline app/code/Magento/Deploy/Console/Command/SetModeCommand.php]: Please add a return after the catch

```
return \Magento\Framework\Console\Cli::RETURN_SUCCESS;
```

**Pull request closed without merging**